nuxeo-runtime/nuxeo-launcher-commons/src/main/java/org/nuxeo/launcher/config/ConfigurationGenerator.java

```diff
@@ -634,12 +642,47 @@ private void logDebugInformation() {
      */
     protected HashMap<String, String> evalDynamicProperties() throws ConfigurationException {
         HashMap<String, String> newParametersToSave = new HashMap<>();
+        evalEnvironmentVariables(newParametersToSave);
         evalLoopbackURL();
         evalServerStatusKey(newParametersToSave);
         return newParametersToSave;
     }
 
     /**
+     * Expand environment variable for properties values of the form ${env:MY_VAR}.
+     *
+     * @since 9.1
+     */
+    protected void evalEnvironmentVariables(Map<String, String> newParametersToSave) {
+        for (Object keyObject : userConfig.keySet()) {
+            String key = (String) keyObject;
+            String value = userConfig.getProperty(key);
+
+            Matcher matcher = ENV_VALUE_PATTERN.matcher(value);
+
+            while (matcher.find()) {
+                boolean booleanValue = "?".equals(matcher.group("boolean"));
+                String envVarName = matcher.group("envparam");
+                String defaultValue = matcher.group("defaultvalue");
+
+                String envValue = getEnvironmentVariableValue(envVarName);
+
+                String result = "";
+                if (booleanValue) {
+                    result = StringUtils.isBlank(envValue) ? "false" : "true";
+                } else {
+                    result = StringUtils.isBlank(envValue) ? defaultValue : envValue;
+                }
+
+                result = matcher.replaceFirst(result);
+                newParametersToSave.put(key, result);
+                matcher = ENV_VALUE_PATTERN.matcher(result);
```

Comment:
Don't recompute the matcher iteratively. Use `Matcher.appendReplacement` in a loop instead. https://docs.oracle.com/javase/8/docs/api/java/util/regex/Matcher.html#appendReplacement-java.lang.StringBuffer-java.lang.String-